### PR TITLE
Fix parsing of password-command option (#6268)

### PR DESCRIPTION
**Pull request by frasertweedale**
The password-command option does not parse its value correctly.
Quotes are ignored, making many kinds of commands impossible to
express (e.g.  `sh -c "foo | bar"`).  Also, `cabal user-config`
treats the argument list as a *list of option values*, rather than a
*value that is a list*.  As a consequence, `cabal user-config
update` corrupts the value in the config file.

Fix these issues by parsing the command as a space separated list of
tokens, and changing the getter to `unwords` the value and return a
*singleton* list.  Also update the argument placeholder from
`PASSWORD` to `COMMAND`.

Fixes: https://github.com/haskell/cabal/issues/6268

## Manual QA Notes

- Please, [see the original bugreport](https://github.com/haskell/cabal/issues/6268#issue-500778440) that contains a short and sweet reproducer.

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.
* [x] Include [manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) if your PR relates to cabal-install.

Bonus points for added automated tests!

### CODE REVIEW

**Comment by ulysses4ever** [comment]
Good job here, thank you! Would it be possible to add a test?

**Comment by frasertweedale** [comment]
@ulysses4ever I added (well, extended) a test, and updated the users guide.

**Comment by frasertweedale** [comment]
*bump* please review when able :)

**Comment by frasertweedale** [comment]
@Kleidukos also affects `cabal upload` command.  Suggest adding label `cabal-install: cmd/upload`.

**Comment by ulysses4ever** [comment]
@frasertweedale please, add a changelog [as described here](https://github.com/haskell/cabal/blob/describe-changelogs-more/CONTRIBUTING.md#changelog).

**Review by ulysses4ever — APPROVED** [review]
Predicated on the changelog, this looks good to me, thanks!

**Comment by ulysses4ever** [comment]
Also, I think it'd be nice to have manual QA for this change. I'll add a note in the OP.

**Comment by frasertweedale** [comment]
@ulysses4ever I added a `changelog.d` entry.

**Comment by ulysses4ever** [comment]
Thank you! Waiting for a second review…

**Review by geekosaur — APPROVED** [review]
Looks reasonable to me

**Comment by frasertweedale** [comment]
Something went wrong with mergify.  Can someone advise on next steps?  I logged into mergify (never used it previously) but it couldn't access the `haskell` dashboard or something...

**Comment by Kleidukos** [comment]
@frasertweedale no problems, thanks

**Comment by jgotoh** [comment]
Here are the results of QA:

1. Verify with latest cabal 3.10.1.0:

`~/.cabal/config` contains: 
`password-command: sh -c "ls | tail -n 1"`

Running `cabal user-config update` results in following change in `~/.cabal/config`:
`password-command: sh,-c,"ls,|,tail,-n,1`

=> password-command was changed, bug exists

2. Verify fix with latest cabal head:

Running `cabal user-config update` results in following change in `~/.cabal/config`:
`password-command: sh -c "ls | tail -n 1"`

Same as before, so QA is passed :1st_place_medal:

**Comment by ulysses4ever** [comment]
@jgotoh thanks a bunch!

**Comment by frasertweedale** [comment]
> hold on, I checked the wrong file, sorry. The following change occurs in my config file after running `cabal-head user-config update` `password-command: "sh" "-c" "ls | tail -n 1"`
> 
> Is this the correct behavior? @frasertweedale

This is expected.  **The values are semantically equivalent.** It's an artifact of the token parsing compared to the pretty-printing of the option values.  There are two ways of representing simple token (with quotes, or without). So unless we represent which variant was used in the option values, there is no way to ensure the string representation round-trips.

**Comment by jgotoh** [comment]
Awesome, thank you very much for your quick answer and your detailed description. Then QA is passed here!  :)

**Comment by ulysses4ever** [comment]
Bugfix -- qualifies for a backport, I think.

**Comment by ulysses4ever** [comment]
@Mergify backport 3.10

**Pull request merged**
